### PR TITLE
[FLOC-4240] GCE driver logging

### PR DESCRIPTION
**Pull request by sarum90**
This PR adds logging for the GCE driver.

What is logged:

- Begin/End and status of all GCE async operations.
- The raw results of list_volumes, listing BlockDeviceVolumes of the volumes in the cluster, and just the name and descriptions of volumes not in the cluster.
- A few other specific messages to log blockdevice_id for the attach and detach operations.

This seems like a lot of logging, but I believe it to be approximately the logging that we currently do for EBS.

I'll compare logs of functional tests before submitting.

### CODE REVIEW

**Comment by sarum90** [comment]
Not super-scientific, but test.log for GCE functional tests with this branch is 463.42KB and test.log for EBS functional tests is 625.07KB at master.

Now there are probably a different number of tests in the EBS functional tests and the GCE functional tests, but at least we are probably in the right ballpark for log size in comparison to EBS.

File: flocker/node/agents/gce.py [hunk 1]
```diff
@@ -172,13 +173,19 @@ def wait_for_operation(compute, operation, timeout_steps):
...
 
-    def finished_operation_result():
-        latest_operation = poller.poll(compute)
-        if latest_operation['status'] == 'DONE':
-            return latest_operation
-        return None
+    with start_action(
+            action_type=u"flocker:node:agents:gce:wait_for_operation",
+            operation=operation
+    ) as action:
```
Comment (justnoise) [inline flocker/node/agents/gce.py]: I would suggest going for 4 space indents for these context manager arguments.  I know it looks better with 8 but, searching for `with start_action` in the rest of the codebase shows that the other folks elected to go with 4 spaces.

**Comment by wallrj** [comment]
Thanks @sarum90 

I looked these GCE functional test logs and the first log Eliot log entries are AWS related. What's up with that?
 * http://ci-live.clusterhq.com:8080/job/ClusterHQ-flocker/job/gce-driver-logging-FLOC-4240/view/storage_drivers/job/_run_trial_for_gce_storage_driver_on_CentOS_7_flocker.node.agents.functional/1/artifact/_trial_temp/test.log
```
3900f920-f2c5-4919-a261-f245defb90de
+-- flocker:node:agents:blockdevice:aws:volume_status_change_wait@1
    |-- needs_attach_data: True
    |-- status: detaching
    |-- target_status: in-use
    |-- timestamp: 2016-03-01 00:54:05.636289
    |-- volume_id: vol-9c48a689
    `-- wait_time: 0

```

File: flocker/node/agents/gce.py [hunk 2]
```diff
@@ -448,44 +488,58 @@ def create_volume(self, dataset_id, size):
...
+                if e.resp.status == 400:
+                    # TODO(mewert): verify with the rest API that this is the
+                    # only way to get a 400.
+                    raise UnknownVolume(blockdevice_id)
+                else:
+                    raise e
+            errors = result.get('error', {}).get('errors', [])
+            for e in errors:
+                if e.get('code') == u"RESOURCE_IN_USE_BY_ANOTHER_RESOURCE":
+                    raise AlreadyAttachedVolume(blockdevice_id)
```
Comment (wallrj) [inline flocker/node/agents/gce.py]: I got hung up on this for a while:

```
c5133747-8a8f-4659-acb6-2e4701edcb5e
+-- flocker:node:agents:gce:attach_volume@1/started
    |-- attach_to: jenkins-centos7-slave-1
    |-- blockdevice_id: flocker-v1-32da4a17-d6ad-4354-891f-d6b98112ad1e
    `-- timestamp: 2016-03-01 00:54:49.635471
    +-- flocker:node:agents:gce:wait_for_operation@2,1/started
        |-- operation:
            |-- id: 11655645756643125264
            |-- insertTime: 2016-02-29T16:54:49.870-08:00
            |-- kind: compute#operation
            |-- name: operation-1456793689649-52cf23661ef68-233b41fc-763f2729
            |-- operationType: attachDisk
            |-- progress: 0
            |-- selfLink: https://www.googleapis.com/compute/v1/projects/clusterhqjenkins/zones/us-central1-f/operations/operation-1456793689649-52cf23661ef68-233b41fc-763f2729
            |-- startTime: 2016-02-29T16:54:50.216-08:00
            |-- status: PENDING
            |-- targetId: 8348751227309162405
            |-- targetLink: https://www.googleapis.com/compute/v1/projects/clusterhqjenkins/zones/us-central1-f/instances/jenkins-centos7-slave-1
            |-- user: 453036982657-compute@developer.gserviceaccount.com
            `-- zone: https://www.googleapis.com/compute/v1/projects/clusterhqjenkins/zones/us-central1-f
        `-- timestamp: 2016-03-01 00:54:50.420111
        +-- flocker:node:agents:gce:wait_for_operation@2,2/succeeded
            |-- final_operation:
                |-- endTime: 2016-02-29T16:54:50.666-08:00
                |-- error:
                    `-- errors: [{u'message': u"The disk resource 'projects/clusterhqjenkins/zones/us-central1-f/disks/flocker-v1-32da4a17-d6ad-4354-891f-d6b98112ad1e' is already being used by 'projects/clusterhqjenkins/zones/us-central1-f/instances/jenkins-centos7-slave-1'", u'code': u'RESOURCE_IN_USE_BY_ANOTHER_RESOURCE'}]
                |-- httpErrorMessage: BAD REQUEST
                |-- httpErrorStatusCode: 400
                |-- id: 11655645756643125264
                |-- insertTime: 2016-02-29T16:54:49.870-08:00
                |-- kind: compute#operation
                |-- name: operation-1456793689649-52cf23661ef68-233b41fc-763f2729
                |-- operationType: attachDisk
                |-- progress: 100
                |-- selfLink: https://www.googleapis.com/compute/v1/projects/clusterhqjenkins/zones/us-central1-f/operations/operation-1456793689649-52cf23661ef68-233b41fc-763f2729
                |-- startTime: 2016-02-29T16:54:50.216-08:00
                |-- status: DONE
                |-- targetId: 8348751227309162405
                |-- targetLink: https://www.googleapis.com/compute/v1/projects/clusterhqjenkins/zones/us-central1-f/instances/jenkins-centos7-slave-1
                |-- user: 453036982657-compute@developer.gserviceaccount.com
                `-- zone: https://www.googleapis.com/compute/v1/projects/clusterhqjenkins/zones/us-central1-f
            `-- timestamp: 2016-03-01 00:54:51.605244
    +-- flocker:node:agents:gce:attach_volume@3/failed
        |-- exception: flocker.node.agents.blockdevice.AlreadyAttachedVolume
        |-- reason: flocker-v1-32da4a17-d6ad-4354-891f-d6b98112ad1e
        `-- timestamp: 2016-03-01 00:54:51.606140

:

```

Couldn't understand how an HTTP error code was being logged and yet the action succeeded.

Is that how the Python library exposes errors in the operation or is that what I'd expect to see if I interacted with the REST API directly?

And it seems strange that the Python API leaks the HTTP error for the initial API request but not for the operations request.

File: flocker/node/agents/gce.py [hunk 3]
```diff
@@ -448,44 +488,58 @@ def create_volume(self, dataset_id, size):
...
-        try:
-            # TODO(mewert): Verify timeout and error conditions.
-            # TODO(mewert): Test what happens when disk is attached RW to a
-            #               different instance, raise the correct error.
-            result = self._do_blocking_operation(
-                self._compute.instances().attachDisk,
-                instance=attach_to,
-                body=config
+            try:
+                # TODO(mewert): Verify timeout and error conditions.
```
Comment (wallrj) [inline flocker/node/agents/gce.py]: :question: I thought I'd already reviewed and approved the realistic GCE timeout branch. Perhaps it's not merged yet.

File: flocker/node/agents/gce.py [hunk 4]
```diff
@@ -194,10 +201,15 @@ def get_metadata_path(path):
...
-                     timeout=timeout_sec)
-    return r.text
+    with start_action(
+            action_type=u"flocker:node:agents:gce:get_metadata_path",
+            path=path
+    ) as action:
+        r = requests.get(_METADATA_SERVER + path,
+                         headers=_METADATA_HEADERS,
+                         timeout=timeout_sec).text
+        action.add_success_fields(response=r)
```
Comment (wallrj) [inline flocker/node/agents/gce.py]: :question: `response = r.text` was clearer.

File: flocker/node/agents/gce.py [hunk 5]
```diff
@@ -402,19 +413,48 @@ def allocation_unit(self):
...
-            BlockDeviceVolume(
-                blockdevice_id=unicode(disk['name']),
-                size=int(GiB(int(disk['sizeGb'])).to_Byte()),
-                attached_to=_extract_attached_to(disk),
-                dataset_id=_blockdevice_id_to_dataset_id(disk['name'])
+        with start_action(
+                action_type=u"flocker:node:agents:gce:list_volumes",
+        ) as action:
+            result = self._compute.disks().list(project=self._project,
+                                                zone=self._zone).execute()
```
Comment (wallrj) [inline flocker/node/agents/gce.py]: :question: I thought I'd seen paging done here.

File: flocker/node/agents/gce.py [hunk 6]
```diff
@@ -402,19 +413,48 @@ def allocation_unit(self):
...
+
+            # 'description' will not even be in the dictionary if no
+            # description was specified.
+            def disk_in_cluster(disk):
+                return (
+                    disk['name'].startswith(_PREFIX) and
+                    (disk.get('description') ==
```
Comment (wallrj) [inline flocker/node/agents/gce.py]: :question: Why change `disk['description'] to `disk.get('description')` ? Don't disks always have descriptions and if not, don't we want to see an exception and a traceback so that we can diagnose the circumstances?

Comment (wallrj) [inline flocker/node/agents/gce.py]: Oh wait, I didn't read the comment above. Ignore me.

Comment (wallrj) [inline flocker/node/agents/gce.py]: Although if the disk has our `_PREFIX` then it would be a bug if it didn't also have a description, right?

File: flocker/node/agents/gce.py [hunk 7]
```diff
@@ -402,19 +413,48 @@ def allocation_unit(self):
...
+                     self._disk_resource_description())
+                )
+
+            Message.log(
+                message_type=u'flocker:node:agents:gce:list_volumes:ignored',
+                ignored_volumes=list({'name': disk['name'],
+                                      'description': disk.get('description')}
+                                     for disk in result['items']
+                                     if not disk_in_cluster(disk))
```
Comment (wallrj) [inline flocker/node/agents/gce.py]: :question: Seems wasteful to check the origins of all  the disks here and again later. Can't we do this in one loop and assign to a `belongs` and an `ignored` list then log ignored at the end.

File: flocker/node/agents/gce.py [hunk 8]
```diff
@@ -448,44 +488,58 @@ def create_volume(self, dataset_id, size):
...
+            try:
+                # TODO(mewert): Verify timeout and error conditions.
+                # TODO(mewert): Test what happens when disk is attached RW to a
+                #               different instance, raise the correct error.
+                result = self._do_blocking_operation(
+                    self._compute.instances().attachDisk,
+                    instance=attach_to,
+                    body=config
+                )
+            except HttpError as e:
```
Comment (wallrj) [inline flocker/node/agents/gce.py]: :question: It might be easier to answer the question below if we always logged a traceback here. If the deployer is working correctly, this shouldn't normally happen, I think.

File: flocker/node/agents/gce.py [hunk 9]
```diff
@@ -448,44 +488,58 @@ def create_volume(self, dataset_id, size):
...
+                if e.resp.status == 400:
+                    # TODO(mewert): verify with the rest API that this is the
+                    # only way to get a 400.
+                    raise UnknownVolume(blockdevice_id)
+                else:
+                    raise e
```
Comment (wallrj) [inline flocker/node/agents/gce.py]: :question: Consider `raise` rather than `raise e` although I see it was like this before.

File: flocker/node/agents/gce.py [hunk 10]
```diff
@@ -501,32 +555,44 @@ def _get_attached_to(self, blockdevice_id):
...
+                    zone=self._zone,
+                    disk=blockdevice_id
+                ).execute()
+            except HttpError as e:
+                if e.resp.status == 404:
+                    # TODO(mewert) Verify with the rest API this is the only
+                    # way to get a 404.
+                    raise UnknownVolume(blockdevice_id)
+                else:
+                    raise e
```
Comment (wallrj) [inline flocker/node/agents/gce.py]: :question: `raise`

File: flocker/node/agents/gce.py [hunk 11]
```diff
@@ -501,32 +555,44 @@ def _get_attached_to(self, blockdevice_id):
...
+            # this point). Might involve putting all GCE interactions behind a
+            # zope interface and then using a proxy implementation to inject
+            # code.
+            self._do_blocking_operation(
+                self._compute.instances().detachDisk, instance=attached_to,
+                deviceName=blockdevice_id)
+            return None
 
     def get_device_path(self, blockdevice_id):
         # TODO(mewert): Verify that we need this extra API call.
```
Comment (wallrj) [inline flocker/node/agents/gce.py]: It's probably only there to satisfy the test for `UnknownVolume`  in `test_get_device_path_unknown_volume`.

**Comment by wallrj** [comment]
Thanks @sarum90 

Looks good to me.

Please 
 * answer or address the comments above,
 * merge forward, 
 * resolve conflicts and 
 * merge when green.

File: flocker/node/agents/gce.py (hunk 5 above)
Comment (justnoise) [inline flocker/node/agents/gce.py]: Paging is still waiting on jenkins executors...

File: flocker/node/agents/gce.py (hunk 6 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: I pose it should not be the case that our code should blow up based on someone outside of flocker rudely deciding to spin up a disk with a malicious name and no description.

I'll add logging to indicate that something fishy is going on in that case, but raising exceptions based on the state of someone's GCE project seems like a bad idea.

File: flocker/node/agents/gce.py (hunk 1 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: Thanks! Fixed.

File: flocker/node/agents/gce.py (hunk 4 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: Done.

I sort of don't like having r.text on 2 lines in a row, but I think I'll just get over that.

File: flocker/node/agents/gce.py (hunk 5 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: I'll be sure to merge forward. I'll even be nice and wait for Brendan to merge first.

File: flocker/node/agents/gce.py (hunk 7 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: Given that disk_in_cluster is a pure function I don't think it's that wasteful, and think that's slightly worse for memory pressure if the size of the lists are approximately equal. (This list can be deleted here and the same memory can be re-used for the next list.)

With lists this small though, readability/maintainability is probably preferred over complexity/size analysis.

I have a vague idea that I've written fewer bugs with this functional-style formation rather than the iterate through and add to one list or the other, but no particular reason to prefer one over the other. Given that you prefer the other I'll go ahead with that formation.

File: flocker/node/agents/gce.py (hunk 3 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: Yup, I'll be sure to resolve the merge.

File: flocker/node/agents/gce.py (hunk 9 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: Fixed

File: flocker/node/agents/gce.py (hunk 8 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: Added within the 400 block. The re-raised exception should be logged by higher layers if they care. The exception that we eat here though should have it's traceback written.

File: flocker/node/agents/gce.py (hunk 2 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: The HTTP error code is confusingly part of the Rest API. It is possible to start an operation, get a 200 response, and keep querying the status of the operation always getting 200 responses, and have the body of one of those responses look like above.

Click "Equivalent REST" on https://console.cloud.google.com/compute/operationsDetail/zones/us-central1-f/operations/operation-1456191905224-52c66193c5541-87cb8805-b0cc7b7c?project=clusterhq-acceptance for an example.

For some reason they decided to use HTTP error codes to describe the result of operations, despite the fact that long Operations are actually GCE REST Nouns, and should probably have some different status states that are not HTTP statuses.

File: flocker/node/agents/gce.py (hunk 11 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: +1, decided to keep it as it happens infrequently enough. And... I might as well try to pass the test we have our external partners pass.

**Comment by sarum90** [comment]
Re: the AWS logs. They are also there for Cinder: http://ci-live.clusterhq.com:8080/job/ClusterHQ-flocker/job/gce-driver-logging-FLOC-4240/view/storage_drivers/job/run_trial_for_cinder_storage_driver_on_CentOS_7_flocker.node.agents.functional/1/artifact/_trial_temp/test.log

So I'm not going to worry too much

File: flocker/node/agents/gce.py [hunk 12]
```diff
@@ -417,18 +436,18 @@ def _do_blocking_operation(self, function, timeout_sec=60, **kwargs):
...
             resource.
         """
-        timeout = kwargs.pop('timeout', VOLUME_DEFAULT_TIMEOUT)
-        sleep = kwargs.pop('sleep', None)
         args = dict(project=self._project, zone=self._zone)
         args.update(kwargs)
         operation = function(**args).execute()
         return wait_for_operation(
-            self._compute, operation, [1]*timeout, sleep)
+            self._compute, operation, [1]*timeout_sec, sleep)
```
Comment (sarum90) [inline flocker/node/agents/gce.py]: @justnoise would you mind glancing at this? I think timeout_sec was unused after something, and so when I merged master in I think I did a drive-by fix of that. Care to double check I did it right?

Comment (justnoise) [inline flocker/node/agents/gce.py]: yep, that actually does _Le Smart_ correctly.

**Comment by sarum90** [comment]
Got green runs of the GCE functional tests on jenkins, so I'm going to use the merge-pr script at this point. Thanks for the reviews!

**Comment by sarum90** [comment]
This PR doesn't touch Cinder code. So I'm going to merge despite the 2 failing cinder functional tests.
